crates/bevy_ecs/src/system/system.rs

```diff
@@ -66,14 +67,16 @@ pub(crate) fn check_system_change_tick(
     change_tick: u32,
     system_name: &str,
 ) {
-    let tick_delta = change_tick.wrapping_sub(*last_change_tick);
-    const MAX_DELTA: u32 = (u32::MAX / 4) * 3;
-    // Clamp to max delta
-    if tick_delta > MAX_DELTA {
+    let delta = change_tick.wrapping_sub(*last_change_tick);
+    // This comparison assumes that `delta` has not overflowed `u32::MAX` before, which will be true
+    // so long as this check always runs before that can happen.
+    if delta > MAX_TICK_DELTA {
+        // TODO: Don't spam this warning over and over for the same system.
         warn!(
-            "Too many intervening systems have run since the last time System '{}' was last run; it may fail to detect changes.",
-            system_name
+            "System '{}' has not run for {} ticks. \
+            Changes older than {} ticks will not be detected.",
+            system_name, delta, MAX_TICK_DELTA - 1,
         );
-        *last_change_tick = change_tick.wrapping_sub(MAX_DELTA);
+        *last_change_tick = change_tick.wrapping_sub(MAX_TICK_DELTA);
```

Comment:
We could change this behavior, and produce false positives instead of false negatives. The current behavior will spam warnings (and relatively checks) every single time an inactive system is checked until it's eventually (if ever) run again. I think that just setting the change tick to the current change tick here is more graceful.

Reply:
(Echoing my sentiment on Discord here) I'd prefer not to alter the change detection behavior in this PR, I just wanted to make it behave consistently in the corner case.

I'm actually not sure what the user in #3071 saw, maybe "warning spam" meant many different systems triggered the warning during one scan. The function that reports that warning should only run during a `check_tick` scan, and every scan has to wait for at least `CHECK_TICK_THRESHOLD` ticks to elapse after the previous scan.

Reply:
Yep, I'd call this comment "resolved" then.